packages/SwingSet/src/kernel/kernel.js

```diff
@@ -660,42 +773,108 @@ export default function buildKernel(
+   * @returns {Promise<CrankResults>}
    */
   async function processUpgradeVat(message) {
     assert(vatAdminRootKref, `initializeKernel did not set vatAdminRootKref`);
     const { vatID, upgradeID, bundleID, vatParameters } = message;
     insistCapData(vatParameters);
 
-    if (!vatWarehouse.lookup(vatID)) {
-      return null; // vat is dead, splat
+    const vatInfo = vatWarehouse.lookup(vatID);
+    if (!vatInfo) {
+      return NO_DELIVERY_CRANK_RESULTS; // vat terminated already
     }
+    const { meterID } = vatInfo;
     const vatKeeper = kernelKeeper.provideVatKeeper(vatID);
     /** @type { import('../types-external.js').KernelDeliveryStopVat } */
     const kd1 = harden(['stopVat']);
     const vd1 = vatWarehouse.kernelDeliveryToVatDelivery(vatID, kd1);
     const status1 = await deliverAndLogToVat(vatID, kd1, vd1);
-    if (status1.terminate) {
-      // TODO: if stopVat fails, stop now, arrange for everything to
-      // be unwound. TODO: we need to notify caller about the failure
-      console.log(`-- upgrade-vat stopVat failed: ${status1.terminate}`);
+
+    function makeFailure(_errorCD) {
+      insistCapData(_errorCD); // capdata(Error)
+      // const args = [upgradeID, false, JSON.parse(errorCD.body)];
+      // actually we shouldn't reveal the details, so instead we do:
+      const error = {
+        '@qclass': 'error',
+        name: 'Error',
+        message: 'vat-upgrade failure',
+      };
+      const args = [upgradeID, false, error];
+      return {
+        body: JSON.stringify(['vatUpgradeCallback', args]),
+        slots: [],
+      };
+    }
+
+    // We use deliveryCrankResults to parse the stopVat status.
+    const results1 = deliveryCrankResults(vatID, status1, false);
+
+    // We don't meter stopVat, since no user code is running, but we
+    // still report computrons to the runPolicy
+    let { computrons } = results1; // BigInt or undefined
+    if (computrons !== undefined) {
+      assert.typeof(computrons, 'bigint');
     }
 
-    // stop the worker, delete the transcript and any snapshot
+    // TODO: if/when we implement vat pause/suspend, and if
+    // deliveryCrankResults changes to not use .terminate to indicate
+    // a problem, this should change to match: where we would normally
+    // pause/suspend a vat for a delivery error, here we want to
+    // unwind the upgrade.
+
+    if (results1.terminate) {
+      // get rid of the worker, so the next delivery to this vat will
+      // re-create one from the previous state
+      await vatWarehouse.killWorker(vatID);
+
+      // notify vat-admin of the failed upgrade
+      const vatAdminMethargs = makeFailure(results1.terminate.info);
+
+      // we still report computrons to the runPolicy
+      const results = harden({
+        ...results1,
+        computrons,
+        abort: true, // always unwind
+        popDelivery: true, // don't repeat the upgrade
+        terminate: undefined, // do *not* terminate the vat
```

Comment:
Yeah, needing to break those cases apart is what drove the separation of `DeliveryStatus` and `CrankResults`: normal `dispatch.deliver` sends want to terminate and re-queue (at least for now, since we have all that splat logic already), but failed upgrade wants to do exactly the opposite.

When we get to separate input queues for each vat, I think any case that terminates the vat will also walk the input queue (in reverse order) and preemptively splat each pending message, so we can delete the input queue at the same time as we delete the vat worker. (unless that represents too much work to do all at once, in which case the mostly-dead vat will exist in a state with a non-empty input queue and no worker, and each time it gets scheduled, we'll splat a few more messages, and then when the input and output queues are finally empty, we can finish deleting it)

Reply:
> I think any case that terminates the vat will also walk the input queue (in reverse order) and preemptively splat each pending message, so we can delete the input queue at the same time as we delete the vat worker.

I actually don't think we'll be able to. An `exit` syscall will terminate the vat, but all syscalls done in the same crank are still valid and should be processed. And we'll have an ordering guarantee that the outbound queue is processed before the inbound queue. So that means we can't splat anything in the inbound queue until we process the outbound queue, and there is no reason a vat termination should accelerate that processing.